client/clientimpl.go

```diff
@@ -0,0 +1,348 @@
+			statusReport.AgentDescription = &protobufs.AgentDescription{
+				AgentType:    w.settings.AgentType,
+				AgentVersion: w.settings.AgentVersion,
+			}
+
+			statusReport.ServerProvidedAllAddonsHash = w.settings.LastServerProvidedAllAddonsHash
+		},
+	)
+
+	w.startConnectAndRun()
+
+	w.isStarted = true
+
+	return nil
+}
+
+func (w *Impl) Stop(ctx context.Context) error {
+	if !w.isStarted {
+		return errors.New("cannot stop because not started")
+	}
+
+	w.isStoppingMutex.Lock()
+	cancelFunc := w.runCancel
+	w.isStoppingFlag = true
+	w.isStoppingMutex.Unlock()
+
+	cancelFunc()
+
+	// Close connection if any.
+	w.connMutex.RLock()
+	conn := w.conn
+	w.connMutex.RUnlock()
+
+	if conn != nil {
+		conn.Close()
+	}
+	// If we are not connected by this point we may still get connected because
+	// we are racing with tryConnectOnce() func. However, after tryConnectOnce we will
+	// check the isStopping flag in receiverLoop() and will exit receiverLoop, so we will
+	// not deadlock.
+
+	// Wait until stopping is finished.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-w.stoppedSignal:
+	}
+	return nil
+}
+
+func (w *Impl) SetAgentAttributes(attrs map[string]protobufs.AnyValue) error {
+	panic("implement me")
+}
+
+func (w *Impl) SetEffectiveConfig(config *protobufs.EffectiveConfig) error {
+	w.sender.UpdateStatus(func(statusReport *protobufs.StatusReport) {
+		statusReport.EffectiveConfig = config
+	})
+	w.sender.ScheduleSend()
+	return nil
+}
+
+// Try to connect once. Returns an error if connection fails and optional retryAfter
+// duration to indicate to the caller to retry after the specified time as instructed
+// by the server.
+func (w *Impl) tryConnectOnce(ctx context.Context) (err error, retryAfter time.Duration) {
+	var resp *http.Response
+	conn, resp, err := w.dialer.DialContext(ctx, w.url.String(), w.requestHeader)
+	if err != nil {
+		if w.settings.Callbacks != nil {
+			w.settings.Callbacks.OnConnectFailed(err)
+		}
+		if resp != nil {
+			w.logger.Errorf("Server responded with status=%v", resp.Status)
+			duration := extractRetryAfterHeader(resp)
+			return err, duration
+		}
+		return err, 0
+	}
+
+	// Successfully connected.
+	w.connMutex.Lock()
+	w.conn = conn
+	w.connMutex.Unlock()
+	if w.settings.Callbacks != nil {
+		w.settings.Callbacks.OnConnect()
+	}
+
+	return nil, 0
+}
+
+// Extract Retry-After response header if the status is 503 or 429. Returns
+// 0 duration if the header is not found or the status is different.
+func extractRetryAfterHeader(resp *http.Response) time.Duration {
+	if resp.StatusCode == http.StatusServiceUnavailable ||
+		resp.StatusCode == http.StatusTooManyRequests {
+		retryAfter := strings.TrimSpace(resp.Header.Get(retryAfterHTTPHeader))
+		if retryAfter != "" {
+			retryIntervalSec, err := strconv.Atoi(retryAfter)
+			if err == nil {
+				retryInterval := time.Duration(retryIntervalSec) * time.Second
+				return retryInterval
+			}
+		}
+	}
+	return 0
```

Comment:
I think that should be fine since `ExponentialBackoff` is used later on, but just wanted to check if that was the intention (I imagine that we could also consider returning the recommended value - 30s - as a default one rather than 0)

Reply:
Refactored a bit to make more explicit that the interval is undefined. The caller takes care of the interval in that case.